### PR TITLE
Subset generator

### PR DESCRIPTION
**Pull request by giorgiosironi**
Related to https://github.com/giorgiosironi/eris/issues/58. Two different generators for:

- generating a set given another `Generator` for its elements (the original `set()`)
- generating the possible subsets of a fixed universe set represented as an `array` (the new `subset()`)

The size is scaled on the maximum, and an index in the list of all possible subsets is chosen by limiting with this scaled size so that `size` effectively limits the subset `count()`. The universe set is very rarely generated.

Sample output:
```
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(1) {
  [0]=>
  int(10)
}
array(0) {
}
array(1) {
  [0]=>
  int(8)
}
array(1) {
  [0]=>
  int(10)
}
array(1) {
  [0]=>
  int(6)
}
array(1) {
  [0]=>
  int(6)
}
array(1) {
  [0]=>
  int(8)
}
array(1) {
  [0]=>
  int(8)
}
array(1) {
  [0]=>
  int(10)
}
array(3) {
  [0]=>
  int(4)
  [1]=>
  int(6)
  [2]=>
  int(8)
}
array(3) {
  [0]=>
  int(6)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(1) {
  [0]=>
  int(10)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(10)
}
array(4) {
  [0]=>
  int(4)
  [1]=>
  int(6)
  [2]=>
  int(8)
  [3]=>
  int(10)
}
array(2) {
  [0]=>
  int(2)
  [1]=>
  int(10)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(10)
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(1) {
  [0]=>
  int(10)
}
array(1) {
  [0]=>
  int(10)
}
array(0) {
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(1) {
  [0]=>
  int(8)
}
array(0) {
}
array(3) {
  [0]=>
  int(6)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(6)
}
array(1) {
  [0]=>
  int(10)
}
array(2) {
  [0]=>
  int(6)
  [1]=>
  int(8)
}
array(2) {
  [0]=>
  int(2)
  [1]=>
  int(10)
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(3) {
  [0]=>
  int(2)
  [1]=>
  int(4)
  [2]=>
  int(10)
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(1) {
  [0]=>
  int(10)
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(0) {
}
array(1) {
  [0]=>
  int(6)
}
array(0) {
}
array(1) {
  [0]=>
  int(8)
}
array(0) {
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(3) {
  [0]=>
  int(4)
  [1]=>
  int(6)
  [2]=>
  int(8)
}
array(3) {
  [0]=>
  int(4)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(1) {
  [0]=>
  int(6)
}
array(3) {
  [0]=>
  int(2)
  [1]=>
  int(6)
  [2]=>
  int(8)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(3) {
  [0]=>
  int(2)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(1) {
  [0]=>
  int(8)
}
array(0) {
}
array(2) {
  [0]=>
  int(6)
  [1]=>
  int(8)
}
array(1) {
  [0]=>
  int(8)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(1) {
  [0]=>
  int(8)
}
array(2) {
  [0]=>
  int(6)
  [1]=>
  int(10)
}
array(3) {
  [0]=>
  int(4)
  [1]=>
  int(6)
  [2]=>
  int(8)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(3) {
  [0]=>
  int(2)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(2) {
  [0]=>
  int(2)
  [1]=>
  int(6)
}
array(2) {
  [0]=>
  int(6)
  [1]=>
  int(8)
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(1) {
  [0]=>
  int(10)
}
array(0) {
}
array(0) {
}
array(0) {
}
array(0) {
}
array(2) {
  [0]=>
  int(8)
  [1]=>
  int(10)
}
array(3) {
  [0]=>
  int(6)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(3) {
  [0]=>
  int(6)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(2) {
  [0]=>
  int(6)
  [1]=>
  int(8)
}
array(3) {
  [0]=>
  int(4)
  [1]=>
  int(6)
  [2]=>
  int(8)
}
array(1) {
  [0]=>
  int(2)
}
array(2) {
  [0]=>
  int(4)
  [1]=>
  int(8)
}
array(3) {
  [0]=>
  int(4)
  [1]=>
  int(6)
  [2]=>
  int(8)
}
array(3) {
  [0]=>
  int(6)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
array(3) {
  [0]=>
  int(2)
  [1]=>
  int(8)
  [2]=>
  int(10)
}
```

### CODE REVIEW

File: src/Eris/Generator/SubsetGenerator.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+class SubsetGenerator implements Generator
+{
+    private $universe;
+    
+    public function __construct(array $universe)
+    {
+        $this->universe = $universe;
+    }
+
+    public function __invoke($size)
```
Comment (giorgiosironi) [inline src/Eris/Generator/SubsetGenerator.php]: we could configure a Generator to require $size (absolute) or $relativeSize (from 0.0 to 1.0), opinions?

File: src/Eris/Generator/SubsetGenerator.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                $subset[] = $this->universe[$i];
+            } 
+        }
+
+        return $subset;
+    }
+
+    public function shrink($set)
+    {
+        // TODO: see SetGenerator::shrink()
```
Comment (giorgiosironi) [inline src/Eris/Generator/SubsetGenerator.php]: going to remove this duplication

File: src/Eris/Generator/SubsetGenerator.php (hunk 1 above)
Comment (MirkoBonadei) [inline src/Eris/Generator/SubsetGenerator.php]: Tell me more about this. How would you implement it?

Comment (giorgiosironi) [inline src/Eris/Generator/SubsetGenerator.php]: optional method `sizeSpecification()` to implement on a Generator with possible values some constants such as `ABSOLUTE`, `RELATIVE`

**Pull request merged**